macros/support/src/class/class_constructor.rs

```diff
@@ -2,48 +2,111 @@ use super::{class::Interface, Class};
 use proc_macro2::TokenStream;
 use quote::quote;
 
-/// Function used to instantiate the COM fields, such as vpointers for the COM object.
+/// Function used to instantiate the COM class
 pub fn generate(class: &Class) -> TokenStream {
     let name = &class.name;
     let vis = &class.visibility;
 
-    // Allocate function signature
-    let allocate_parameters = &class.fields;
+    let parameters = &class.fields;
+    let user_fields = class.fields.iter().map(|f| {
+        let name = &f.ident;
+        quote! {
+            #name: ::std::mem::ManuallyDrop::new(#name)
+        }
+    });
 
     let interface_inits = gen_vpointer_inits(class);
+    let ref_count_ident = crate::utils::ref_count_ident();
 
-    // Syntax for instantiating the fields of the struct.
     let interfaces = &class.interfaces;
     let interface_fields = gen_allocate_interface_fields(interfaces);
-    let ref_count_field = gen_allocate_ref_count_field();
-    let user_fields = gen_allocate_user_fields(class);
 
     quote! {
-        #vis fn new(#(#allocate_parameters),*) -> #name {
+        /// Allocate the COM class
+        #vis fn allocate<T: ::com::Interface>(#(#parameters),*) -> Option<T> {
             #interface_inits
-
-            #name {
+            let instance = #name {
                 #interface_fields
-                #ref_count_field
-                #user_fields
-            }
+                #ref_count_ident: std::cell::Cell::new(1),
+                #(#user_fields),*
+            };
+            let instance = ::std::mem::ManuallyDrop::new(::std::boxed::Box::pin(instance));
+            instance.query()
         }
     }
 }
 
-// User field input as parameters to the allocate function.
-fn gen_allocate_user_fields(class: &Class) -> TokenStream {
-    let field_idents = class.fields.iter().map(|f| &f.ident);
+/// Function used to instantiate a default version COM class
```

Comment:
Not sure what this comment is trying to tell me.

Reply:
Whoops. This is a function used to create an instance of the class where all the user fields are default. We don't want to provide a normal `Default` implementation because that would allow the user to stack allocate the class which we don't want.